### PR TITLE
Solve #164, make http.py return ordered headers.

### PR DESCRIPTION
**Pull request by sunhao2014**
Solve #164, make http.py return ordered headers, also add the related unit test and setup dependencies.

BTW, according to Kiran's suggestion, some invalid http header test cases were added.

### CODE REVIEW

**Comment by sunhao2014** [comment]
Just noticed that the PR failed the auto build.
image
The reason why py26 didn't work is we used the `collections.OrderedDict` which is supported by py27. Thus how do we deal with this problem? Should we add support to Python versions older than 2.7?

**Comment by kbandla** [comment]
We could have, if ordereddict was a _necessity_. However:
* many users probably dont need the exact header order
* python2.6 is still widely used

How about doing this instead:
`try` to import collections. On `ImportError`, use a standard python dict. If not, use `collections.OrderedDict`

**Comment by brifordwylie** [comment]
We could pull https://pypi.python.org/pypi/ordereddict as a dependency. One of the nice things about dpkt is we have no requirements outside the standard library, seems a shame to add a dependency just for this small thing.

**Comment by brifordwylie** [comment]
Just saw Kiran's comment, yeah that's the right thing to do,

File: dpkt/http.py [hunk 1]
```diff
@@ -4,11 +4,12 @@
 
 import cStringIO
 import dpkt
+import collections
```
Comment (kbandla) [inline dpkt/http.py]: try:
    from collections import OrderedDict
except ImportError:
    OrderedDict = dict

File: dpkt/http.py [hunk 2]
```diff
@@ -4,11 +4,12 @@
...
 
 
 def parse_headers(f):
     """Return dict of HTTP headers parsed from a file object."""
-    d = {}
+    d = collections.OrderedDict()  # keep the order of the header.
```
Comment (kbandla) [inline dpkt/http.py]: d = OrderedDict()

**Comment by Saylenty** [comment]
Guys, is it really usefull feature? If we use ordered dict we would lose in speed of execution (and I think even in seconds). Once, I tried to optimise one problem (https://github.com/Saylenty/algorithm_with_dots-matrices) trying to solve it using ordered set, but this was not a good idea.

**Comment by obormot** [comment]
@Saylenty +1

**Comment by kbandla** [comment]
Well, it is true too. OrderedDict is written in python (vs C). 

What if we do this instead? (This will not break any existing code either)
```
parse_headers(f, ordered=False):
    d = {}
    if ordered:
        from collections import OrderedDict
        d = OrderedDict()
   ....
```

**Comment by obormot** [comment]
@kbandla I like this option...  but tbh my favorite solution is to not implement this at all :-)   the HTTP protocol spec does not dictate any order in the headers, so dpkt is solid there. If someone wants them in order for their specific use case then it's easy to write their own parse function and feed it tcp.data.

**Comment by kbandla** [comment]
:+1: @obormot lets not implement this. go ahead and remove it @sunhao2014 (ordereddict support)

**Comment by sunhao2014** [comment]
Sure. I'll close this PR and send a new PR that only contains the added http invalid test cases.

**Comment by Saylenty** [comment]
@kbandla , Personally, i would rather implement this feature using OrderedDIct as you've written and write a warning only in the documentation that it can be much slower...

**Pull request closed without merging**